modules/javafx.graphics/src/main/java/com/sun/javafx/scene/TreeShowingExpression.java

```diff
@@ -0,0 +1,150 @@
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.sun.javafx.scene;
+
+import com.sun.javafx.binding.ExpressionHelper;
+import javafx.beans.InvalidationListener;
+import javafx.beans.binding.BooleanExpression;
+import javafx.beans.value.ChangeListener;
+import javafx.scene.Node;
+import javafx.scene.Scene;
+import javafx.stage.Window;
+
+/**
+ * Used to observe changes in tree showing status for a {@link Node}.  For a Node's tree to be showing
+ * it must be visible, its ancestors must be visible, the node must be part of a {@link Scene} and
+ * the scene must have a {@link Window} which is currently showing.<p>
+ *
+ * This class provides the exact same functionality as {@link NodeHelper#isTreeShowing(Node)} in
+ * an observable form.
+ */
+public class TreeShowingExpression extends BooleanExpression {
+    private final ChangeListener<Boolean> windowShowingChangedListener = (obs, oldVal, newVal) -> updateTreeShowing();
+
+    private final ChangeListener<Window> sceneWindowChangedListener = (scene, oldWindow, newWindow) -> {
+        if (oldWindow != null) {
+            oldWindow.showingProperty().removeListener(windowShowingChangedListener);
+        }
+        if (newWindow != null) {
+            newWindow.showingProperty().addListener(windowShowingChangedListener);
+        }
+        updateTreeShowing();
+    };
+
+    private final ChangeListener<Scene> nodeSceneChangedListener = (node, oldScene, newScene) -> {
+        if (oldScene != null) {
+            oldScene.windowProperty().removeListener(sceneWindowChangedListener);
+        }
+        if (newScene != null) {
+            newScene.windowProperty().addListener(sceneWindowChangedListener);
+        }
+
+        sceneWindowChangedListener.changed(
+            null,
+            oldScene == null ? null : oldScene.getWindow(),
+            newScene == null ? null : newScene.getWindow()
+        );
+    };
+
+    private final Node node;
+
+    private ExpressionHelper<Boolean> helper;
+    private boolean valid;
+    private boolean treeShowing;
+
+    /**
+     * Constructs a new instance.
+     *
+     * @param node a {@link Node} for which the tree showing status should be observed, cannot be null
+     */
+    public TreeShowingExpression(Node node) {
+        this.node = node;
+        this.node.sceneProperty().addListener(nodeSceneChangedListener);
+
+        NodeHelper.treeVisibleProperty(node).addListener(windowShowingChangedListener);
+
+        nodeSceneChangedListener.changed(null, null, node.getScene());  // registers listeners on Window/Showing
```

Comment:
We do not follow this pattern of calling the changed() or invalidated() methods. Instead we directly add the necessary calls at the time of initialisation and disposal. I think the pattern should be followed here too.
For reference, the pattern is followed in all Skin classes. For example [here](https://github.com/openjdk/jfx/blob/887a443693a04d29ffaff8b8da353db839a987b4/modules/javafx.controls/src/main/java/javafx/scene/control/skin/ButtonSkin.java#L148) in ButtonSkin, we do not make a call to `sceneChangeListener.changed()`, instead the necessary calls are directly made on [next lines](https://github.com/openjdk/jfx/blob/887a443693a04d29ffaff8b8da353db839a987b4/modules/javafx.controls/src/main/java/javafx/scene/control/skin/ButtonSkin.java#L151) and then [opposite](https://github.com/openjdk/jfx/blob/887a443693a04d29ffaff8b8da353db839a987b4/modules/javafx.controls/src/main/java/javafx/scene/control/skin/ButtonSkin.java#L179) calls in dispose().

For this line 89 needs to be replaced by following code, (with additional null checks)
```
node.getScene().windowProperty().addListener(sceneWindowChangedListener);
node.getScene().getWindow().showingProperty().addListener(windowShowingChangedListener);
updateTreeShowing();
```

and similarly they should be removed in dispose().

Reply:
Isn't it quite error prone to repeat this logic again (especially with all the null cases), not to mention that you would need to test the code for the initial case (with/without Scene, with/without Window), the "in use" case and again for the disposal case?

I personally use helpers for this kind of property chaining as it is far to easy to get wrong:

    public Binding<Boolean> isShowing(Node node) {
        Values.of(node.sceneProperty())
           .flatMap(s -> Values.of(s.windowProperty()))
           .flatMap(w -> Values.of(w.showingProperty()))
           .orElse(false)
           .toBinding();
    }

The implementation here takes care of `null` and automatically tracks property changes and is type safe to boot.  I think JavaFX in general could really benefit from this, as I've seen this chaining logic repeated a lot.

Reply:
My concern is about having a similar way of doing something. It would keep the code uniform. We have been following the earlier pattern under a cleanup task [JDK-8241364](https://bugs.openjdk.java.net/browse/JDK-8241364). Several bugs under this task are being fixed in earlier way.
May be we can discuss the new way of handling properties under a separate issue and plan to modify all such instances at once. Does that sound ok ?

Reply:
hmm ... might appear convenient (in very controlled contexts) but looks like a precondition violation: the sender of the change must not be null (concededly not explicitly spec'ed but logically implied, IMO)

so would tend to _not_ see this as blueprint for a general pattern fx code base

Reply:
So, what you are suggesting is to replace the above single line to this:

        Scene scene = node.getScene();
        
        if (scene != null) {
            Window window = scene.getWindow();
            
            scene.addListener(nodeSceneChangedListener);
            
            if (window != null) {
                window.showingProperty().addListener(windowShowingChangedListener);
            }
        }

        updateTreeShowing();

And something similarly wordy in dispose again -- and donot forget that I would then need to also duplicate the testing code to make sure all these branches are covered in both the constructor and in dispose again.

Are you sure?   I could extract the methods instead if you donot like me passing in `null` for the `ObservableValue`, it would then look like:

        sceneChanged(null, node.getScene());  // register chain

        sceneChanged(node.getScene(), null);  // unregister chain

No further changes in testing required as it is all covered and proven to work correctly.

Also I should mention that this is not Skin code in case it was missed.

Reply:
I could also split the register/unregister part into separate methods for Scene and Window (as you can see, I have a great aversion against duplicating code, mostly because it would then need to be tested multiple times as well).

Reply:
I took a quick look at your latest change, and it seems reasonable to me now that you are calling a helper method rather than triggering a change method on the listener. I'll take a closer look later in the week. In the mean time @kleopatra and @arapte can provide their feedback.

Reply:
not going for a full review - just a comment: agree with Kevin, the delegate methods are the way out, basically look good

`No further changes in testing required as it is all covered` - a bold statement .. looks like there's a missed null check (which was in the listener code but didn't make it into the delegate) at the end of sceneChanged

    windowChanged(oldScene.getWindow(), newScene.getWindow());

any of old/new scene can be null (or what am I missing?) If that's not covered by the tests .. ;)

Reply:
> any of old/new scene can be null (or what am I missing?) If that's not covered by the tests .. ;)

Yep. And if you look at the checks run by GitHub actions, there are failing tests. For example:

```
2021-02-09T20:57:51.7835652Z test.com.sun.javafx.scene.control.infrastructure.ControlSkinFactoryTest > testAlternativeSkinAssignable FAILED
2021-02-09T20:57:51.7838200Z     java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
2021-02-09T20:57:51.7841382Z         at test.com.sun.javafx.scene.control.infrastructure.ControlSkinFactory.createAlternativeSkin(ControlSkinFactory.java:323)
2021-02-09T20:57:51.7845300Z         at test.com.sun.javafx.scene.control.infrastructure.ControlSkinFactory.replaceSkin(ControlSkinFactory.java:302)
2021-02-09T20:57:51.7852081Z         at test.com.sun.javafx.scene.control.infrastructure.ControlSkinFactoryTest.testAlternativeSkinAssignable(ControlSkinFactoryTest.java:91)
2021-02-09T20:57:51.7854816Z 
2021-02-09T20:57:51.7855105Z         Caused by:
2021-02-09T20:57:51.7855858Z         java.lang.reflect.InvocationTargetException
2021-02-09T20:57:51.7857628Z             at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
2021-02-09T20:57:51.7860264Z             at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:64)
2021-02-09T20:57:51.7863798Z             at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
2021-02-09T20:57:51.7866625Z             at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:500)
2021-02-09T20:57:51.7868214Z             at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:481)
2021-02-09T20:57:51.7870982Z             at test.com.sun.javafx.scene.control.infrastructure.ControlSkinFactory.createAlternativeSkin(ControlSkinFactory.java:321)
2021-02-09T20:57:51.7873192Z             ... 2 more
2021-02-09T20:57:51.7873407Z 
2021-02-09T20:57:51.7873689Z             Caused by:
2021-02-09T20:57:51.7874236Z             java.lang.NullPointerException
2021-02-09T20:57:51.7875765Z                 at javafx.graphics/com.sun.javafx.scene.TreeShowingExpression.sceneChanged(TreeShowingExpression.java:127)
2021-02-09T20:57:51.7877830Z                 at javafx.graphics/com.sun.javafx.scene.TreeShowingExpression.<init>(TreeShowingExpression.java:66)
2021-02-09T20:57:51.7879836Z                 at javafx.controls/javafx.scene.control.skin.ProgressIndicatorSkin.<init>(ProgressIndicatorSkin.java:130)
2021-02-09T20:57:51.7881678Z                 at javafx.controls/javafx.scene.control.skin.ProgressBarSkin.<init>(ProgressBarSkin.java:97)
2021-02-09T20:57:51.7883928Z                 at test.com.sun.javafx.scene.control.infrastructure.ControlSkinFactory$ProgressBarSkin1.<init>(ControlSkinFactory.java:515)
2021-02-09T20:57:51.7885496Z                 ... 8 more
```

Reply:
good ;)

Reply:
somehow missed that the red cross on the checks below might be related to test failures (vs. pre-checking compile issues or such only ;) Any way to see the failing tests only? Scrolling through the complete log is a bit .. well

Reply:
Sadly, there is no way to view just the failures -- at least not right now. What I do is open the raw log and look for "FAILED" (all caps).

I'll file an RFE to see whether we can produce a summary of the test failures, but I won't have time to look at it right away.

Reply:
thanks for the info :)

Reply:
> I could extract the methods instead if you donot like me passing in `null` for the `ObservableValue`, it would then look like:
> 
> ```
>     sceneChanged(null, node.getScene());  // register chain
> 
>     sceneChanged(node.getScene(), null);  // unregister chain
> ```
> 

That's a good pattern, IMO (biased, using it in my own code)

> 
> Also I should mention that this is not Skin code in case it was missed.

No, it was not missed :)  Aside: actually, we cannot use the pattern above in skin code because we are using LambdaMultipleListenerHandler to un/register for change notification. The consumer that's registered carries only the observable, consequently its old value is not available at notification time. Which forces keeping the old value stored in the skin, if interested in changes from a "path" property.

Reply:
I saw the failures in github actions, but figured it was an intermittent failure as I couldn't see an obvious test failure.  Somehow the local build didn't run the test again, so I missed it completely.  I've pushed a fix.

I'm happy this pattern is more to your liking, it does look a bit cleaner and saves some duplication.